### PR TITLE
A small diagnostic improvement for dropping_copy_types

### LINKED ISSUE #125189: Confusing suggestion for `dropping_copy_types`
### Code

I had the following code:

```rust
    let w_fd = w.as_raw_fd();
    let cmd_stdout = unsafe { Stdio::from_raw_fd(w_fd) };
    let cmd_stderr = unsafe { Stdio::from_raw_fd(w_fd) };
    // Not required, but prevents us from accidentally using it later.
    drop(w_fd);
```

if we compile this, `rustc` rightly warns us:

```
warning: calls to `std::mem::drop` with a value that implements `Copy` does nothing
   --> src/main.rs:158:5
    |
158 |     drop(w_fd);
    |     ^^^^^----^
    |          |
    |          argument has type `i32`
    |
    = note: use `let _ = ...` to ignore the expression or result
    = note: `#[warn(dropping_copy_types)]` on by default
```

This bit, however:

```
    = note: use `let _ = ...` to ignore the expression or result
```

… I'm not clear on. Let's take the compiler's suggestion and modify the code:

```rust
    // Not required, but prevents us from accidentally using it later.
    let _ = drop(w_fd);
```

Now we'll get an even more confusing diagnostic:

```
warning: calls to `std::mem::drop` with a value that implements `Copy` does nothing
   --> src/main.rs:158:13
    |
158 |     let _ = drop(w_fd);
    |             ^^^^^----^
    |                  |
    |                  argument has type `i32`
    |
    = note: use `let _ = ...` to ignore the expression or result
    = note: `#[warn(dropping_copy_types)]` on by default
```

…we're now `note`'d to do something we're already doing.

A competing interpretation of this diagnostic, in my mind, is that it means to do `let _ = <the original expression where we compute w_fd>`. But we cannot do that, as we use `w_fd`. I suppose we could inline all the uses, but I wonder if `rustc` shouldn't realize that no, the variable must be named / that suggestion breaks the compile. Or, should `rustc` understand that we cannot `let _ =` this variable, and that we'd need to inline it? But here there are two inlines, and in the general case, side-effects in the expression could make this tricky.

The higher level thing here is that I was attempting to drop a variable from scope, essentially, by moving it with `drop`. Obviously, that doesn't work for `Copy` types. I guess I could use `{}` to scope `w_fd` … but unfortunately I need both `cmd_stdout` and `cmd_stderr` later.

### Current output

```Shell
warning: calls to `std::mem::drop` with a value that implements `Copy` does nothing
   --> src/main.rs:158:5
    |
158 |     drop(w_fd);
    |     ^^^^^----^
    |          |
    |          argument has type `i32`
    |
    = note: use `let _ = ...` to ignore the expression or result
    = note: `#[warn(dropping_copy_types)]` on by default
```

### Desired output

_No response_

### Rationale and extra context

_No response_

### Other cases

_No response_

### Rust Version

```Shell
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: x86_64-unknown-linux-gnu
release: 1.76.0
LLVM version: 17.0.6
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by surechen**
For a value `m`  which implements `Copy` trait, `drop(m);` does nothing. 
We now suggest user to ignore it by a abstract and general note: `let _ = ...`.
I think we can give a clearer note here: `let _ = m;`

fixes #125189

### CODE REVIEW

File: compiler/rustc_lint/messages.ftl
Comment (compiler-errors) [inline compiler/rustc_lint/messages.ftl:199]: Why not just make this a `#[suggestion]`?

Comment (Urgau) [reply compiler/rustc_lint/messages.ftl:199]: It would need to be `maybe-incorrect`, because we don't want users to blindly apply the suggestion without first checking their code for a possible logic error.

File: tests/ui/associated-types/defaults-unsound-62211-1.next.stderr
Comment (Urgau) [inline tests/ui/associated-types/defaults-unsound-62211-1.next.stderr:9]: `let _ = expr;` still drops the expression directly.

File: compiler/rustc_lint/messages.ftl
Comment (compiler-errors) [reply compiler/rustc_lint/messages.ftl:199]: Of course, but `maybe-incorrect` exists exactly for this reason. I'd rather we not manually implement a "suggestion-like" note when we have the infrastructure to render suggestions.

File: tests/ui/associated-types/defaults-unsound-62211-1.next.stderr
Comment (surechen) [reply tests/ui/associated-types/defaults-unsound-62211-1.next.stderr]: > `let _ = expr;` still drops the expression directly.

Done. Thank you.

File: tests/ui/lint/dropping_copy_types.stderr
Comment (Urgau) [inline tests/ui/lint/dropping_copy_types.stderr:29]: Generally we try to make the suggestion explicit about what the it does, this is particularly relevant in IDEs were they might not display the change at all but just the suggestion message.

What about simply reusing the "note" message?
```suggestion
help: use `let _ = ...` to ignore the expression or result
```

File: compiler/rustc_lint/src/lints.rs
Comment (Urgau) [inline compiler/rustc_lint/src/lints.rs:679]: I would prefer if we not print the note and the suggestion at the same time. Something like this should do it.

```suggestion
    #[subdiagnostic]
    pub sugg: DropCopySuggestion,
}

#[derive(Subdiagnostic)]
pub enum DropCopySuggestion {
    #[note(lint_dropping_copy_types_note)]
    Note {
        #[label]
        label: Span,
    }
    #[suggestion(lint_dropping_copy_types_suggestion, style = "verbose", code = "{replace}", applicability = "maybe-incorrect")]
    Suggestion {
        #[primary_span]
        sugg: Span,
        replace: String,
    }
}
```

File: tests/ui/lint/dropping_copy_types-issue-125189.rs
Comment (Urgau) [inline tests/ui/lint/dropping_copy_types-issue-125189.rs:1]: It would be great if we could test the applicability of the suggestion with `rustfix`. This can be achieved with `run-rustfix` and having no others warnings/errors (ie move the match in a different UI test file).

```suggestion
//@ check-fail
//@ run-rustfix
```

File: compiler/rustc_lint/messages.ftl
Comment (surechen) [reply compiler/rustc_lint/messages.ftl]: Done. Thank you.

File: tests/ui/lint/dropping_copy_types.stderr
Comment (surechen) [reply tests/ui/lint/dropping_copy_types.stderr]: Done. Thank you.

File: compiler/rustc_lint/src/lints.rs
Comment (surechen) [reply compiler/rustc_lint/src/lints.rs]: Done. Thank you.

File: tests/ui/lint/dropping_copy_types-issue-125189.rs
Comment (surechen) [reply tests/ui/lint/dropping_copy_types-issue-125189.rs:1]: Done. Thank you.

File: compiler/rustc_lint/src/lints.rs
Comment (Urgau) [inline compiler/rustc_lint/src/lints.rs:694]:
```suggestion
    #[multipart_suggestion(
        lint_suggestion,
        style = "verbose",
        applicability = "maybe-incorrect"
    )]
    Suggestion {
        #[suggestion_part(code = "let _ = ")]
        start_span: Span,
        #[suggestion_part(code = "")]
        end_span: Span,
    },
```

File: tests/ui/lint/dropping_copy_types-issue-125189-can-not-fixed.rs
Comment (Urgau) [inline tests/ui/lint/dropping_copy_types-issue-125189-can-not-fixed.rs:14]: Nit: these 2 lines don't really fit in with the name of the test, since they can be fixed automatically

File: compiler/rustc_lint/src/drop_forget_useless.rs
Comment (surechen) [reply compiler/rustc_lint/src/drop_forget_useless.rs]: > I don't think we need to use `span_to_snippet` here, we can take the [Span](https://doc.rust-lang.org/beta/nightly-rustc/rustc_span/struct.Span.html#method.shrink_to_lo) of the expression and the arg and do some span arithmetic to avoid it. My idea is to get the span of `drop(` and the span of `)` and in the same suggestion replace the first part by `let _ = ` and the second by nothing.
> 
> That gives better output, since the diagnostic machinery can print more targeted diff.

Your suggestions are  much better, thank you very much.

File: compiler/rustc_lint/src/lints.rs
Comment (surechen) [reply compiler/rustc_lint/src/lints.rs]: Done. Thank you.

**Review by Urgau — APPROVED** [review]
Looks good to me! Thanks.

*(you will have to wait for @compiler-errors for the final approval)*

**Comment by Urgau** [comment]
Btw, the same improvement can also be done for the `dropping_references`, `forgetting_copy_types` and `forgetting_types` lints, they have the same note, which should also be a proper suggestion. (feel free to do them in this PR as separate commits or as a follow-up PR if you like, no obligation to do so).

**Comment by surechen** [comment]
> Btw, the same improvement can also be done for the `dropping_references`, `forgetting_copy_types` and `forgetting_references` lints, they have the same note, which should also be a proper suggestion. (feel free to do them in this PR as separate commits or as a follow-up PR if you like, no obligation to do so).

Thank you very much, I tend to implement in follow-up PRs.

**Comment by surechen** [comment]
Hi, Thanks to the committers who are working hard to merge this PR.
I submitted a PR(https://github.com/rust-lang/cargo/pull/13964) to rust-lang/cargo to try to fix the CI problem. 
I think  we can try again after that PR is merged into rust-lang/rust/src/tools/cargo

**Comment by Urgau** [comment]
https://github.com/rust-lang/cargo/pull/13964 was merged in https://github.com/rust-lang/rust/pull/125682. Let's retry merging this PR.

@bors r=Urgau

**Pull request merged**
